Pax.cs

```diff
@@ -187,49 +210,87 @@ public class Frontend {
 
-      // FIXME currently assuming that the assembly only contains stuff that concerns
-      //       us. it would be best to filter by the implemented interfaces, and focus on
-      //       interfaces made available by Pax.
-      foreach (Type ty in PaxConfig.assembly.GetExportedTypes())
+      // Inspect each type that implements PacketProcessor, trying to instantiate it for use
+      foreach (Type type in PaxConfig.assembly.GetExportedTypes()
+                                            .Where(typeof(PacketProcessor).IsAssignableFrom))
       {
-#if MOREDEBUG
-        Console.WriteLine("Trying to instantiate {0}", ty);
-#endif
-        PacketProcessor pp = (PacketProcessor)Activator.CreateInstance(ty); // NOTE this means that we require "ty" to define a default constructor.
-
         // Find which network interfaces this class is handling
         List<int> subscribed = new List<int>();
-
+        PacketProcessor pp = null;
         for (int idx = 0; idx < PaxConfig.no_interfaces; idx++)
         {
+          // Does this interface have this type specified as the lead handler?
           if ((!String.IsNullOrEmpty(PaxConfig.interface_lead_handler[idx])) &&
-              ty.Name == PaxConfig.interface_lead_handler[idx])
+              type.Name == PaxConfig.interface_lead_handler[idx])
           {
+            // Only instatiate pp if needed
+            if (pp == null)
+              pp = InstatiatePacketProcessor(type);
+            if (pp == null)
+              // If pp is still null, then we couldn't instantiate it.
+              break;
             subscribed.Add(idx);
             PaxConfig.interface_lead_handler_obj[idx] = pp;
           }
         }
 
+        // Print which interfaces this type is the handler for
         var tmp = Console.ForegroundColor;
         Console.Write (indent);
         Console.ForegroundColor = ConsoleColor.Green;
-        Console.Write (ty);
+        Console.Write (type);
         Console.ForegroundColor = ConsoleColor.Gray;
         Console.Write (" <- ");
         Console.ForegroundColor = ConsoleColor.Yellow;
         Console.WriteLine(
            String.Join(", ", subscribed.ConvertAll<string>(ofs => PaxConfig.deviceMap[ofs].Name)));
-        // FIXME could also print type-related info of ty, such as which Pax interfaces it implements.
+
+        // List the Pax interfaces this type implements:
+        Console.ForegroundColor = ConsoleColor.Gray;
+        Console.Write("  : ");
+        Console.ForegroundColor = ConsoleColor.Cyan;
+        Console.WriteLine(String.Join(", ", PacketProcessorHelper.GetUsedPaxTypes(type).Select(t => t.Name)));
         Console.ForegroundColor = tmp;
       }
       // FIXME add check to see if there's an interface that references a lead_handler that doesn't appear in the assembly. That should be flagged up to the user, and lead to termination of Pax.
-
+    }
+    private static PacketProcessor InstatiatePacketProcessor(Type type)
+    {
+      #if MOREDEBUG
```

Comment:
Not sure how usual this is, but "usually" pragma lines are not indented (at least, the # isn't). This helps them stand out.